sdk/samples/iot/hub/src/pnp_helper.c

```diff
@@ -0,0 +1,438 @@
+  AZ_RETURN_IF_FAILED(az_json_writer_append_int32(&json_writer, ack_code));
+  AZ_RETURN_IF_FAILED(
+      az_json_writer_append_property_name(&json_writer, desired_temp_ack_version_name));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_int32(&json_writer, ack_version));
+
+  if (az_span_ptr(ack_description) != NULL)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_json_writer_append_property_name(&json_writer, desired_temp_ack_description_name));
+    AZ_RETURN_IF_FAILED(az_json_writer_append_string(&json_writer, ack_description));
+  }
+
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+  AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+
+  if (az_span_ptr(component_name) != NULL)
+  {
+    AZ_RETURN_IF_FAILED(az_json_writer_append_end_object(&json_writer));
+  }
+
+  *out_span = az_json_writer_get_bytes_used_in_destination(&json_writer);
+
+  return result;
+}
+
+// Process the twin properties and invoke user callback for each property
+az_result pnp_helper_process_twin_data(
+    az_json_reader* json_reader,
+    bool is_partial,
+    const az_span** sample_components_ptr,
+    int32_t sample_components_num,
+    char* scratch_buf,
+    int32_t scratch_buf_len,
+    pnp_helper_property_callback property_callback,
+    void* context_ptr)
+{
+  az_json_reader copy_json_reader;
+  int32_t version;
+  int32_t len;
+  int32_t index;
+
+  AZ_RETURN_IF_FAILED(az_json_reader_next_token(json_reader));
+
+  if (!is_partial
+      && (sample_json_child_token_move(json_reader, sample_iot_hub_twin_desired) != AZ_OK))
+  {
+    printf("Failed to get desired property\r\n");
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  copy_json_reader = *json_reader;
+  if ((sample_json_child_token_move(&copy_json_reader, sample_iot_hub_twin_desired_version)
+       != AZ_OK)
+      || az_failed(az_json_token_get_int32(&(copy_json_reader.token), (int32_t*)&version)))
+  {
+    printf("Failed to get version\r\n");
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  while (az_succeeded(az_json_reader_next_token(json_reader)))
+  {
+    if (json_reader->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+    {
+      if (az_failed(az_json_token_get_string(
+              &(json_reader->token), (char*)scratch_buf, (int32_t)scratch_buf_len, (int32_t*)&len)))
+      {
+        printf("Failed to string value for property name\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if ((len == (int32_t)az_span_size(sample_iot_hub_twin_desired_version))
+          && (memcmp(
+                  (void*)az_span_ptr(sample_iot_hub_twin_desired_version),
+                  (void*)scratch_buf,
+                  (size_t)az_span_size(sample_iot_hub_twin_desired_version))
+              == 0))
+      {
+        continue;
+      }
```

Comment:
You should not do direct memcmp comparing byte-by-byte for property name look-ups within the JSON.
Use the built-in comparison API: `az_json_token_is_text_equal`

Other than correctness, there is an added benefit that you don't need to copy every token into your scratch buffer, even if they would be ignored.

Reply:
Alright changed it to check with `az_json_token_get_string` first. If matches, it advances and `continue`. Else it `_get_string`'s